vscode/src/diagnostics.ts

```diff
@@ -130,3 +140,151 @@ export function updateQSharpJsonDiagnostics(
+}
+
+export function clearCommandDiagnostics() {
+  if (!commandDiagnostics) {
+    log.warn(`no diagnostic collection for commands, not clearing`);
+  }
+
+  commandDiagnostics?.clear();
+}
+
+/**
+ * Wrapper that invokes a function. This function would be something that calls into
+ * the Q# compiler. If the function throws an exception that contains Q# diagnostics,
+ * reports the diagnostics to VS Code before rethrowing the original exception.
+ */
+export async function invokeAndReportCommandDiagnostics<T>(
+  fn: () => T | Promise<T>,
+): Promise<T> {
+  try {
+    // Clear the diagnostics from the last command run.
+    clearCommandDiagnostics();
+    return await fn();
+  } catch (e: unknown) {
+    reportIfQSharpErrors(e);
+    throw e;
+  }
+}
+
+/**
+ * Given an exception, checks if it's a JSON string representation of IQSharpError[],
+ * and if so, reports the diagnostics to VS Code.
+ *
+ * @param e an exception originating from the qsharp-lang package
+ */
+function reportIfQSharpErrors(e: unknown) {
+  if (!commandDiagnostics) {
+    log.warn(`diagnostic collection for commands was not initialized`);
+    return;
+  }
+
+  let qsharpErrors: IQSharpError[] | undefined;
+  if (typeof e === "string") {
+    try {
+      const errors = JSON.parse(e);
+      // Check for the shape of IQSharpError[]
+      if (Array.isArray(errors) && errors.length > 0 && errors[0].document) {
+        qsharpErrors = errors;
+      }
+    } catch (_) {
+      // Couldn't parse the error as JSON.
+    }
+  }
+
+  if (qsharpErrors) {
+    const byUri = new Map<vscode.Uri, vscode.Diagnostic[]>();
+
+    for (const error of qsharpErrors) {
+      const uri = getSourceUri(error.document);
+
+      const diagnostics = byUri.get(uri) || [];
+      error.diagnostic.message = `Q# command error: ${error.diagnostic.message}`;
+      diagnostics.push(toVsCodeDiagnostic(error.diagnostic));
+      byUri.set(uri, diagnostics);
+    }
+
+    for (const [uri, diags] of byUri) {
+      commandDiagnostics.set(uri, diags);
+    }
+
+    // Focus on Problems view
+    vscode.commands.executeCommand("workbench.action.problems.focus");
+
+    vscode.window.showErrorMessage(
+      "The Q# command returned errors. Please see the Problems view.",
+      { modal: true },
+    );
+  }
+}
+
+/**
+ * This is temporary until we're able to report proper stdlib and project URIs from
+ * the wasm layer. See https://github.com/microsoft/qsharp/blob/f8d344b32a1f1f918f3c91edf58c975db10f4370/wasm/src/diagnostic.rs
+ *
+ * @param maybeUri A source name returned from a Q# diagnostic
+ * @returns A VS code URI that's okay to use in a Diagnostic object
+ */
+function getSourceUri(maybeUri: string): vscode.Uri {
+  if (maybeUri === "<project>") {
+    return vscode.Uri.parse(maybeUri, false); // strict=false ignores lack of scheme
+  }
+
+  try {
+    return vscode.Uri.parse(maybeUri, true);
+  } catch (e) {
+    // Not a URI, assume it's a filename from the stdlib
```

Comment:
Maybe not urgent for this PR, but `Not a URI, assume it's a filename from the stdlib` seems dangerous. Are we sure no other errors might get here? Is there some other condition we could check to be sure? (e.g. would it contain `library/std` in the path in that case?)

Reply:
Definitely dangerous. _Right now_ I'm sure, by virtue of having written the code myself, but it's delicate for sure. This would be part of the broader improvements I'd like to make in error handling.

There's pretty much no meaningful indicator we can check. "arrays.qs" is an example of what can end up here.

Reply:
Will address this in a future change.